### PR TITLE
A/B-Tests for network latency and block performance

### PR DESCRIPTION
**Pull request by roypat**
## Changes

Introduces A/B-compatible performance tests for block device throughput and network latency. Also contains minor QoL refractors, such as:
- Outputting correct units in test_ab.py (instead of declaring everything to be milliseconds), including logic to convert un-readable values such as "987654 Bytes" into "987.7KB".
- Introduce consistent naming for variables in test_ab.py, in hopes of making the code easier to read/maintain
- Generalize test_snapshot_ab.py's `TestSetup.dimensions` property to be a property of `Microvm` that returns a default set of dimensions shared across all perf tests. 

## Reason

More A/B-Testings and code cleanups

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: tests/integration_tests/performance/test_block_ab.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        .with_arg("--cpus_allowed_policy=split")
+        .with_arg("--randrepeat=0")
+        .with_arg(f"--runtime={RUNTIME_SEC}")
+        .with_arg(f"--write_bw_log={mode}")
+        .with_arg("--log_avg_msec=1000")
+        .with_arg("--output-format=json+")
+        .build()
+    )
+
+    logs_path = Path(microvm.jailer.chroot_base_with_id()) / "fio_output"
```
Comment (JonathanWoollett-Light) [inline tests/integration_tests/performance/test_block_ab.py:77]: Could you add a comment to why the construction of this is changed from https://github.com/firecracker-microvm/firecracker/blob/main/tests/integration_tests/performance/test_block_performance.py#L78C71-L78C73

File: tests/integration_tests/performance/test_block_ab.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        rc, _, stderr = microvm.ssh.run("rm /tmp/*.log")
+        assert rc == 0, stderr
+
+        return logs_path, cpu_load_future.result()
+
+
+def process_fio_logs(vm, fio_mode, logs_dir, metrics):
+    """Parses the fio logs in `{logs_dir}/{fio_mode}_bw.*.log and emits their contents as CloudWatch metrics"""
+    for job_id in range(vm.vcpus_count):
+        data = (logs_dir / f"{fio_mode}_bw.{job_id + 1}.log").read_text("UTF-8")
```
Comment (JonathanWoollett-Light) [inline tests/integration_tests/performance/test_block_ab.py:110]:
```suggestion
        data = Path(f"{logs_dir}/{fio_mode}_bw.{job_id + 1}.log").read_text("UTF-8")
```

File: tests/integration_tests/performance/test_block_ab.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    process_fio_logs(vm, fio_mode, logs_dir, metrics)
+
+    for cpu_util_data_point in list(cpu_load["firecracker"].values())[0]:
+        metrics.put_metric("cpu_utilization_vmm", cpu_util_data_point, "Percent")
+
+    metrics.set_dimensions(
+        {
+            "instance": global_props.instance,
+            "cpu_model": global_props.cpu_model,
+            "host_kernel": "linux-" + global_props.host_linux_version,
```
Comment (JonathanWoollett-Light) [inline tests/integration_tests/performance/test_block_ab.py]:
```suggestion
            "host_kernel": f"linux-{global_props.host_linux_version}",
```

File: tests/framework/microvm.py [hunk 4]
```diff
@@ -360,6 +360,19 @@ def pid_in_new_ns(self):
         # Read the PID stored inside the file.
         return int(pid_file_path.read_text(encoding="ascii"))
 
+    @property
+    def dimensions(self):
+        """Gets a default set of cloudwatch dimensions describing the configuration of this microvm"""
+        return {
+            "instance": global_props.instance,
+            "cpu_model": global_props.cpu_model,
+            "host_kernel": "linux-" + global_props.host_linux_version,
```
Comment (JonathanWoollett-Light) [inline tests/framework/microvm.py:369]: See https://github.com/firecracker-microvm/firecracker/pull/4133#discussion_r1338847497

File: tests/framework/microvm.py [hunk 5]
```diff
@@ -360,6 +360,19 @@ def pid_in_new_ns(self):
...
+            "guest_kernel": self.kernel_file.stem[2:],
+            "rootfs": self.rootfs_file.name,
+            "vcpus": str(self.vcpus_count),
+            "guest_memory": f"{self.mem_size_bytes / 1024 / 1024}MB",
```
Comment (JonathanWoollett-Light) [inline tests/framework/microvm.py:373]:
```suggestion
           "guest_memory": f"{self.mem_size_bytes / 1024**2}MB",
```

**Review by JonathanWoollett-Light — COMMENTED** [review]
Nit: Change the commit message from

> fix: convert numpy.int64 to float in ab_test.py 

to

> fix: convert `numpy.int64` to float in `ab_test.py`

**Review by JonathanWoollett-Light — COMMENTED** [review]
Change

> test: Have ab_test.py output correct units

to

> test: Have `ab_test.py` output correct units

File: tests/host_tools/metrics.py [hunk 6]
```diff
@@ -110,3 +110,89 @@ def emit_raw_emf(emf_msg: dict):
...
+    # due to precision problems (e.g. infinite loop of dividing/multiplying by 1000, alternating
+    # between values < 1 and >= 1000).
+    while abs(value) < 1 and unit in INV_UNIT_REDUCTIONS:
+        value *= 1000
+        unit = INV_UNIT_REDUCTIONS[unit]
+    while abs(value) >= 1000 and unit in UNIT_REDUCTIONS:
+        value /= 1000
+        unit = UNIT_REDUCTIONS[unit]
+
+    return value, unit
```
Comment (JonathanWoollett-Light) [inline tests/host_tools/metrics.py:185]: Would `reduce_value(1_900_000, Bytes)` return `(1, Megabytes)`?

**Review by JonathanWoollett-Light — APPROVED** [review]

File: tests/integration_tests/performance/test_block_ab.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        )
+        # Instruct fio to pin one worker per vcpu
+        .with_arg("--cpus_allowed_policy=split")
+        .with_arg("--randrepeat=0")
+        .with_arg(f"--runtime={RUNTIME_SEC}")
+        .with_arg(f"--write_bw_log={mode}")
+        .with_arg("--log_avg_msec=1000")
+        .with_arg("--output-format=json+")
+        .build()
+    )
```
Comment (JonathanWoollett-Light) [inline tests/integration_tests/performance/test_block_ab.py:75]: Nit: Could be in this PR in its own commit or could be separate PR.

Could we add an `flags` optional parameter to `CmdBuilder` `__init__` like:

```python
class CmdBuilder:
    """Command builder class."""

    def __init__(self, bin_path, flags=[]):
        """Initialize the command builder."""
        self._bin_path = bin_path
        self._args = {flag: "" for flag in flags}
```

this would support

```suggestion
    cmd = (
        CmdBuilder("fio", flags = [
            f"--name={mode}-{block_size}",
            f"--rw={mode}",
            f"--bs={block_size}",
            "--filename=/dev/vdb",
            "--time_base=1",
            f"--size={BLOCK_DEVICE_SIZE_MB}M",
            "--direct=1",
            "--ioengine=libaio",
            "--iodepth=32",
            f"--ramp_time={WARMUP_SEC}",
            f"--numjobs={microvm.vcpus_count}",
            # Set affinity of the entire fio process to a set of vCPUs equal in size to number of workers
            f"--cpus_allowed={','.join(str(i) for i in range(microvm.vcpus_count))}",
            # Instruct fio to pin one worker per vcpu
            "--cpus_allowed_policy=split",
            "--randrepeat=0",
            f"--runtime={RUNTIME_SEC}",
            f"--write_bw_log={mode}",
            "--log_avg_msec=1000",
            "--output-format=json+"
        ])
        .build()
    )
```

File: tests/host_tools/metrics.py [hunk 8]
```diff
@@ -110,3 +110,89 @@ def emit_raw_emf(emf_msg: dict):
...
+    """
+    # Could do this recursively, but I am worried about infinite recursion
+    # due to precision problems (e.g. infinite loop of dividing/multiplying by 1000, alternating
+    # between values < 1 and >= 1000).
+    while abs(value) < 1 and unit in INV_UNIT_REDUCTIONS:
+        value *= 1000
+        unit = INV_UNIT_REDUCTIONS[unit]
+    while abs(value) >= 1000 and unit in UNIT_REDUCTIONS:
+        value /= 1000
+        unit = UNIT_REDUCTIONS[unit]
```
Comment (JonathanWoollett-Light) [inline tests/host_tools/metrics.py:183]: Nit: A little more succinct and maybe slightly more efficient, but arguably less readable. Feel free to ignore.
```suggestion
    while value // 1000 and unit in UNIT_REDUCTIONS:
        value /= 1000
        unit = UNIT_REDUCTIONS[unit]
```

**Review by JonathanWoollett-Light — APPROVED** [review]

**Review by zulinx86 — APPROVED** [review]
